<a href="https://colab.research.google.com/github/somyav950-spec/60-days-coding-challenge/blob/main/Day_4_Sentence_Embeddings.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [5]:
!pip install -U sentence-transformers

In [12]:
from sentence_transformers import SentenceTransformer
from sklearn.metrics.pairwise import cosine_similarity

In [13]:
model = SentenceTransformer('all-MiniLM-L6-v2')

modules.json:   0%|          | 0.00/349 [00:00<?, ?B/s]

config_sentence_transformers.json:   0%|          | 0.00/116 [00:00<?, ?B/s]

README.md:   0%|          | 0.00/10.5k [00:00<?, ?B/s]

sentence_bert_config.json:   0%|          | 0.00/53.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/612 [00:00<?, ?B/s]

model.safetensors: reconstructing file:   0%|          |  0.00B / 90.9MB            

model.safetensors: downloading bytes:           |  0.00B            

Loading weights:   0%|          | 0/103 [00:00<?, ?it/s]

tokenizer_config.json:   0%|          | 0.00/350 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/466k [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/112 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/190 [00:00<?, ?B/s]

In [14]:
sentences = [
    "I love learning artificial intelligence.",
    "I enjoy studying machine learning.",
    "Python is my favorite programming language.",
    "The weather is very pleasant today.",
    "I like eating pizza with my friends."
]

print(sentences)

['I love learning artificial intelligence.', 'I enjoy studying machine learning.', 'Python is my favorite programming language.', 'The weather is very pleasant today.', 'I like eating pizza with my friends.']


In [15]:
embeddings = model.encode(sentences)

print("Number of sentences:", len(sentences))
print("Embedding shape:", embeddings.shape)

Number of sentences: 5
Embedding shape: (5, 384)


In [16]:
similarity_matrix = cosine_similarity(embeddings)

print(similarity_matrix)

[[ 1.0000000e+00  7.3183036e-01  5.1218247e-01 -9.8254532e-04
   2.8077349e-01]
 [ 7.3183036e-01  9.9999988e-01  4.5191759e-01  2.4015466e-02
   2.5609869e-01]
 [ 5.1218247e-01  4.5191759e-01  1.0000000e+00  8.8139981e-02
   2.5902805e-01]
 [-9.8254532e-04  2.4015466e-02  8.8139981e-02  9.9999994e-01
   1.2456299e-01]
 [ 2.8077349e-01  2.5609869e-01  2.5902805e-01  1.2456299e-01
   1.0000000e+00]]


In [17]:
for i in range(len(sentences)):
    for j in range(i + 1, len(sentences)):
        score = similarity_matrix[i][j]
        print(f"Sentence {i+1} & Sentence {j+1}: {score:.4f}")

Sentence 1 & Sentence 2: 0.7318
Sentence 1 & Sentence 3: 0.5122
Sentence 1 & Sentence 4: -0.0010
Sentence 1 & Sentence 5: 0.2808
Sentence 2 & Sentence 3: 0.4519
Sentence 2 & Sentence 4: 0.0240
Sentence 2 & Sentence 5: 0.2561
Sentence 3 & Sentence 4: 0.0881
Sentence 3 & Sentence 5: 0.2590
Sentence 4 & Sentence 5: 0.1246


In [18]:
max_score = -1
best_pair = None

for i in range(len(sentences)):
    for j in range(i + 1, len(sentences)):
        score = similarity_matrix[i][j]

        if score > max_score:
            max_score = score
            best_pair = (i, j)

print("Most similar sentences:")
print(sentences[best_pair[0]])
print(sentences[best_pair[1]])
print("Similarity Score:", round(max_score, 4))

Most similar sentences:
I love learning artificial intelligence.
I enjoy studying machine learning.
Similarity Score: 0.7318


Reflection: I learned that sentence embeddings capture the meaning of sentences instead of just matching words. The AI and Machine Learning sentences received the highest similarity score of 0.7318. I was surprised that AI can understand semantic similarity even when the exact words are different.